python: add bindings for libflux-idset,hostlist

grondo opened the pull request:
This PR is pretty straightforward. It adds basic `IDset` and `Hostlist` Python classes which wrap libflux-idset.so and libflux-hostlist.so.

I was also going to make automated bindings for librlist to make an Rv1 class, however I decided to stop here and make the Rv1 class a separate PR.

grondo commented:
Force-pushed fixes for the python errors caught by LGTM.

garlick commented:
Just playing with this a little.  I can't seem make `hostlist.decode()` raise an exception on bad input.  Instead I get a zero length hostlist.  Is that expected?

grondo commented:
No, and I thought I had tests for that but I must have accidentally dropped them at some point.

Nice catch. I'll work on a fix.

chu11 commented on src/bindings/python/flux/hostlist.py:
```diff
@@ -167,6 +201,6 @@ def copy(self):
 
-def decode(string):
+def decode(arg):
     """Decode a string in RFC 29 hostlist format to a Hostlist object"""
```
should this comment be updated to string or iterable?

grondo replied:
Yeah, thanks. I'm actually not sure of Python idioms, if `hostlist.decode()` just directly calls `hostlist.Hostlist()` is it even useful to separately provide that function?

chu11 replied:
good point.  can't speak for whats common, but since there is an `encode`, i think it's nice to have a `decode`, minimally so people can search for it.

grondo replied:
Ah good point! It doesn't do any harm at least...

grondo commented:
Ok, pushed some fixup commits to address comments and improve interface and tests:

 - Fixed missing exceptions on `hostlist_decode(3)` failures, added tests
 - Improved `Hostlist()` constructor, now takes any valid RFC 29 hostlist string or iterable of said strings
 - Improved `hostlist.append()` to take Hostlist, or iterable or plain string of RFC 29 strings
 - add a few more tests

garlick reviewed (approved):
Nice!  That fixed my problem and offhand I'm not finding anything else.
It's really cool to be able to use these "natively" in python!

I'll give my approval although someone with stronger python-fu like @chu11 probably ought to give a nod also.

grondo commented:
Thanks. As I noted above in a comment, I'm not sure if the separate `hostlist.decode()` and `idset.decode()` functions are the "pythonic" way to do it, or if people expect to just use the class constructors anyway.

I was going to stop here for now and then see if an extra interfaces or minor interface changes popped out of work in Python-based flux utilities like `flux-resource.py`, and add those at that time.

grondo commented:
Ok, perhaps I'll squash the fixup commits unless there is any objection?

chu11 reviewed (approved):
LGTM, but my python-fu is probably not much better than @garlick

grondo commented:
Actually, I'm not sure the approach used here (idset and hostlist each under their own ffi instance will work here.
The `Rv1` or similar implementation will need an ffi instance that includes both the idset and hostlist instances, and I'm not exactly sure how to do that, but I wonder if what we really want here is to build *one* cffi source for all of hostlist, idset, and rlist, or even just put these directly into the `_core.c` ffi? 

I really don't know anything about cffi yet, so I wonder if either @trws or @SteVwonder can give a hint on what might be the right approach here. Note that the use case is that an `Rv1` class would have methods to return `Hostlist` or `IDset` objects directly from `Rv1.nodelist` or `Rv1.ranks` methods or properties. (These methods would call the underlying `rlist_nodelist(struct rlist *rl)` call and then instantiate a Hostlist with something like `Hostlist(handle=result)`

Right now when I attempt that I get the error:

```
TypeError: initializer for ctype 'struct hostlist *' appears indeed to be 'struct hostlist *', but the types are different (check that you are not e.g. mixing up different ffi instances)
```

grondo commented:
Ah, just after the call I found that using `ffi.include()` "seems" to work, e.g. in the new `_idset_build.py` the `hostlist` and `idset` FFIs are included via

```python
from _hostlist_build import ffi as hostlist_ffi
from _idset_build import ffi as idset_ffi

ffi = FFI()

ffi.include(hostlist_ffi)
ffi.include(idset_ffi)

```

Then I am able to instantiate `Hostlist` and `IDset` objects in the `rlist.py` module with simply:

```python
   def nodelist(self):
        return Hostlist(handle=self.pimpl.nodelist())

    def ranks(self):
        return IDset(handle=self.pimpl.ranks())
```

grondo commented:
Ok, since there is a path forward for composing FFI objects from multiple sources, I think this one is ok to go in as is. We can add minor tweaks as we come up with users of these new classes.

I added one more commit here. I noticed the asan builder was timing out when it seemed like the build would have completed given more time. Perhaps with the addition of extra tests the 40m timeout was not quite enough, so I've added a commit to increase the timeout to 50m.

garlick commented:
Sounds good!  Set MWP?

trws commented:
This is exactly what you want here I think.  If they’re included the 
ffi objects and type-identifiers are cross-compatible.  We used to do 
this for several flux core components before re-working the build to 
build all of the standard core components as one module.

On 16 Nov 2020, at 14:29, Mark Grondona wrote:

> Ah, just after the call I found that using `ffi.include()` "seems" to 
> work, e.g. in the new `_idset_build.py` the `hostlist` and `idset` 
> FFIs are included via
>
> ```python
> from _hostlist_build import ffi as hostlist_ffi
> from _idset_build import ffi as idset_ffi
>
> ffi = FFI()
>
> ffi.include(hostlist_ffi)
> ffi.include(idset_ffi)
>
> ```
>
> Then I am able to instantiate `Hostlist` and `IDset` objects in the 
> `rlist.py` module with simply:
>
> ```python
>    def nodelist(self):
>         return Hostlist(handle=self.pimpl.nodelist())
>
>     def ranks(self):
>         return IDset(handle=self.pimpl.ranks())
> ```
>
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/3341#issuecomment-728368672

SteVwonder commented on src/bindings/python/flux/hostlist.py:
```diff
@@ -0,0 +1,208 @@
+            raise IndexError("Hostlist index out of range")
+
+        if isinstance(index, slice):
+            return [self[n] for n in range(len(self))[index]]
```
Minor nitpick:
```suggestion
            return [self[n] for n in range(index.start, index.stop, index.step)]
```
will avoid constructing unnecessary elements if the slice starts at index > 0.

Sidenote: it's kind of annoying that `range` doesn't support taking a slice as an argument.  And that slices don't support `*` expansion.

grondo replied:
@SteVwonder, I did try this approach, but it isn't so simple.

For a slice, I found that any of `start`, `stop`, or `step` can be `None` (or maybe only two of them, I don't remember), and `range` does not support `None` for some arguments. It ended being a bit of mess trying to handle all the corner cases, and `range(len(self))[index]` seemed to be a lot cleaner.

However, your comment made think, why don't we just make a function in `hostlist.py` that will operate like `range()`, but _does_ support slices? I'll give that a try.

trws replied:
itertools.islice?

```python
[i for i in it.islice(range(10), s.start, s.stop, s.step)]
```

grondo replied:
Thanks! It would have taken me awhile to find that one.

grondo replied:
Hm `islice` doesn't support negative values so this won't work:
```python
>>> hl = Hostlist("fluke[0-7]")
>>> hl[::-1]
['fluke7', 'fluke6', 'fluke5', 'fluke4', 'fluke3', 'fluke2', 'fluke1', 'fluke0']
```

with `islice`:
```python
>>> hl = Hostlist("fluke[0-7]")
>>> hl[::-1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/g/g0/grondo/git/flux-core.git/src/bindings/python/flux/hostlist.py", line 134, in __getitem__
    return [i for i in islice(self, index.start, index.stop, index.step)]
ValueError: Step for islice() must be a positive integer or None.

```

trws replied:
Wow, dang.  I thought that was supposed to give you the full slice setup. 😞

trws replied:
Is it just the step that will not take negatives? If so, one last crack:

```python
it = self
if s.step < 0:
  it = reversed(self)
[i for i in it.islice(it, s.start, s.stop, abs(s.step))]
```

Now mainly doing this because this feels like it shouldn't be hard, and yet... and it's bugging me. 😆

grondo replied:
This simple wrapper for `range()` seems to work:
```python
def rangeslice(index, maxi):
    step = index.step
    if step is None:
        step = 1
    start = index.start
    if start is None:
        start = 0 if step > 0 else maxi
    stop = index.stop
    if stop is None:
        stop = maxi + 1 if step > 0 else -1
    return range(start, stop, step)
```

grondo replied:
Oh sorry @trws, Github didn't refresh the page so I missed your suggestion above. I'll give that a try.

trws replied:
Your solution works perfectly nicely as far as I'm concerned, it feels so odd to need to work around this.

grondo replied:
> Is it just the step that will not take negatives? If so, one last crack:

None of the values can take negatives according to the  [docs](https://docs.python.org/3.8/library/itertools.html#itertools.islice):

> Unlike regular slicing, islice() does not support negative values for start, stop, or step.

SteVwonder commented on src/bindings/python/flux/hostlist.py:
```diff
@@ -0,0 +1,208 @@
+        self.pimpl.uniq()
+        return self
+
+    def expand(self):
+        """Convert a Hostlist to a Python list"""
+        return self[:]
```
You can potentially leave this method out.  Because `HostList` implements `__iter__`, users can just do `list(my_hostlist)`.

SteVwonder reviewed (approved):
Thanks @grondo for putting this together. LGTM! Two totally optional comments below:

grondo commented:
Ok, I've pushed an update that replaces the use of `range()[index]` with a new `rangeslice()` function, and also removed the unnecessary `expand` method as suggested by @SteVwonder.

SteVwonder commented:
LGTM! Ready for the MWP label?

grondo commented:
Ok, sure, thanks!

grondo commented:
Ah, this will have to be manually merged since I've modified `.github/workflows`

Let me rebase first.

SteVwonder commented:
All the tests pass, so I'm going to push the button.  Thanks @grondo!

grondo commented:
Thanks!!
